build_tools/doctool/src/com/google/doctool/custom/JavaEmulSummaryDoclet.java

```diff
@@ -0,0 +1,273 @@
+                pw.println("<dl>");
+
+                String packURL = JAVADOC_URL + pack.getQualifiedName().toString()
+                        .replace(".", "/") + "/";
+
+                Iterator<? extends Element> classesIterator = pack.getEnclosedElements()
+                        .stream()
+                        .filter(element -> env.isSelected(element) && env.isIncluded(element))
+                        .filter(element -> element.getModifiers().contains(Modifier.PUBLIC))
+                        .sorted(Comparator.comparing((Element o) -> o.getSimpleName().toString()))
+                        .iterator();
+
+                while (classesIterator.hasNext()) {
+                    Element cls = classesIterator.next();
+                    // Each class links to Oracle's main JavaDoc
+                    emitClassDocs(env, pw, packURL, cls);
+                    if (classesIterator.hasNext()) {
+                        pw.print("\n");
+                    }
+                }
+
+                pw.println("</dl>\n");
+            });
+
+            pw.close();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return true;
+    }
+
+    private void emitClassDocs(DocletEnvironment env, PrintWriter pw, String packURL, Element cls) {
+        pw.format("  <dt><a href=\"%s%s.html\">%s</a></dt>\n", packURL,
+                qualifiedSimpleName(cls), qualifiedSimpleName(cls));
+
+        // Print out all fields
+        String fields = cls.getEnclosedElements()
+                .stream()
+                .filter(element -> element.getKind().isField())
+                .filter(field -> field.getModifiers().contains(Modifier.PUBLIC))
+                .map(field -> field.getSimpleName().toString())
+                .collect(Collectors.joining(", "));
+
+        if (!fields.isEmpty()) {
+            pw.format("  <dd style='margin-bottom: 0.5em;'>%s</dd>\n", fields);
+        }
+
+        List<String> constructors = cls.getEnclosedElements()
+                .stream()
+                .filter(element -> ElementKind.CONSTRUCTOR == element.getKind())
+                .filter(member -> member.getModifiers().contains(Modifier.PUBLIC))
+                .map(member -> (ExecutableElement) member)
+                .map(executableElement -> flatSignature(env, cls, executableElement))
+                .collect(Collectors.toList());
+
+        List<String> methods = cls.getEnclosedElements()
+                .stream()
+                .filter(element -> ElementKind.METHOD == element.getKind())
+                .filter(member -> member.getModifiers().contains(Modifier.PUBLIC))
+                .map(member -> (ExecutableElement) member)
+                .map(executableElement -> flatSignature(env, cls, executableElement))
+                .collect(Collectors.toList());
+
+        List<String> members = new ArrayList<>(constructors);
+        members.addAll(methods);
```

Comment:
It _seems_ better to just put all three of these kinds of members into one big stream... I see that to keep the diff exactly the same, the fields have to be set up on their own.

Fields and constructors though are almost identical, it looks like the difference is basically to just sort constructors before methods?

I think it is probably best leaving alone, since we want to keep the diff the same and just replace the code - use this as the way to ensure it is stable (absent any other tests, etc). Maybe if we have a reason to change this in the future, we could consolidate, or maybe even style things differently.

Reply:
My goal here was to make sure that the output will remain exactly identical to the original one, there was some implicit sorting and ordering in the original API, this way I was able to compare and make sure I didn't change something I shouldn't or I didn't produce a different output.